src/tests/Common/Directory.Build.targets

```diff
@@ -18,6 +18,8 @@
     <CopyCoreDisToolsToCoreRoot Condition="$(GCStressDependsOnCoreDisTools) And '$(DotNetBuildFromSource)' != 'true'">true</CopyCoreDisToolsToCoreRoot>
     <!-- Non-desktop OS's use a custom dotnet host, instead of corerun -->
     <IsDesktopOS Condition="'$(TargetsBrowser)' != 'true' and '$(TargetsAndroid)' != 'true' and '$(TargetstvOS)' != 'true' and '$(TargetsiOS)' != 'true' and '$(TargetsMacCatalyst)' != 'true'">true</IsDesktopOS>
+    <Crossgen2Supported Condition="'$(RuntimeFlavor)' == 'coreclr' and '$(TestBuildMode)' != 'nativeaot'">true</Crossgen2Supported>
+    <UsePublishedCrossgen2 Condition="'$(Crossgen2Supported)' == 'true' and '$(Configuration)' != 'checked'">true</UsePublishedCrossgen2>
```

Comment:
I'm somewhat worried that by forking Crossgen2 behavior for checked vs. release mode we're making life harder for developers. According to the tracking issue it's bad behavior that we're testing our internal raw Crossgen2 build, not the final NativeAOT-compiled thing we ship to customers and I don't see any apparent reason why we should maintain it in checked build mode. (And what about debug mode?) Also, as you can see, the <code>coreclr-runtime outerloop</code> pipeline we're using as one of the main vehicles for Crossgen2 testing uses <code>Checked</code> build mode for all its R2R jobs so your change wouldn't actually translate into any "fidelity improvements" in it. We'd either need to be smarter about running some R2R jobs using checked and others using release mode but for now I'd be inclined to just remove the configuration check. We might want to only set <code>UsePublishedCrossgen2</code> when not explicitly set previously, that would let developers explicitly override it for the purpose of their local innerloop; we could then also create a new occasionally running pipeline (e.g. once a week) that would force-set the property to false so that we could keep some amount of regression testing for the "raw" Crossgen2.

Reply:
/cc @jkoritzinsky and @hoyosjs for additional thoughts on this

Reply:
I switched to other things, but this is still work-in-progress as some tests were failing. 🥲

The only reason why I added this condition was because NativeAOT'd published crossgen2 was failing some asserts. I wanted to (temporarily) get by that error to find out if there are other blockers. Unfortunately, there still are issues even in release and debug builds. Things get interesting when we run the published crossgen2 executable (AOT, singlefile or even a simple self-contained app); `crossgen2.exe <args>` fails where previous `dotnet crossgen2.dll <args>` was working.

Reply:
If you have a hard time figuring out the problem with NativeAOT Crossgen2 we can probably loop in @MichalStrehovsky for help but I'd be surprised if it was completely non-functional as that's literally what we ship :-).

Reply:
Can you please quote an example of Crossgen2 command line and the error it's reporting?

Reply:
Example from new CI run (now `checked` is using `PublishAot`): https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-80154-merge-7538eff8dd144b4eb2/PayloadGroup0/1/console.a9f514c2.log?helixlogtype=result
```
  Starting:    readytorun.tests.XUnitWrapper (parallel test collections = on, max threads = 6)
    readytorun/tests/mainv1/mainv1.sh [FAIL]
      Unhandled Exception: EETypeRva:0x0121FF08(System.Diagnostics.DebugProvider+DebugAssertException): GetFileSystemType returned 26
         at System.Diagnostics.DebugProvider.Fail(String, String) + 0x35
         at System.Diagnostics.Debug.Fail(String, String) + 0x24
         at Interop.Sys.TryGetFileSystemType(SafeFileHandle, Interop.Sys.UnixFileSystemTypes&) + 0xdb
         at Microsoft.Win32.SafeHandles.SafeFileHandle.CanLockTheFile(Interop.Sys.LockOperations, FileAccess) + 0x5b
         at Microsoft.Win32.SafeHandles.SafeFileHandle.Init(String, FileMode, FileAccess, FileShare, FileOptions, Int64, Int64&, UnixFileMode&) + 0x157
         at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String, FileMode, FileAccess, FileShare, FileOptions, Int64, UnixFileMode, Int64&, UnixFileMode&, Func`4) + 0x89
         at System.IO.Strategies.OSFileStreamStrategy..ctor(String, FileMode, FileAccess, FileShare, FileOptions, Int64, Nullable`1) + 0x7f
         at System.IO.FileStream..ctor(String, FileMode, FileAccess, FileShare, Int32, FileOptions, Int64) + 0x9a
         at System.IO.StreamWriter.ValidateArgsAndOpenPath(String, Boolean, Encoding, Int32) + 0x7e
         at ILCompiler.PEWriter.MapFileBuilder.SaveMap(String) + 0x5b
         at ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitPortableExecutable() + 0xb77
         at ILCompiler.ReadyToRunCodegenCompilation.Compile(String) + 0x1c2
         at ILCompiler.Program.RunSingleCompilation(Dictionary`2, InstructionSetSupport, String, Dictionary`2, HashSet`1, ReadyToRunCompilerContext) + 0x1af0
         at ILCompiler.Program.Run() + 0xea4
         at ILCompiler.Crossgen2RootCommand.<>c__DisplayClass187_0.<.ctor>b__0(InvocationContext context) + 0x25b
         at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext) + 0x1e
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext() + 0x105
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<BuildInvocationChain>b__0(InvocationContext invocationContext, Func`2 _) + 0x35
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseErrorReporting>b__0>d.MoveNext() + 0x9b
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<UseParseErrorReporting>b__0(InvocationContext context, Func`2 next) + 0x39
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseHelp>b__0>d.MoveNext() + 0x4e
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<UseHelp>b__0(InvocationContext context, Func`2 next) + 0x39
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext() + 0x168
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<UseVersionOption>b__0(InvocationContext context, Func`2 next) + 0x39
         at System.CommandLine.Invocation.InvocationPipeline.<Invoke>g__FullInvocationChain|3_0(InvocationContext) + 0x7c
         at System.CommandLine.Invocation.InvocationPipeline.Invoke(IConsole) + 0x81
         at ILCompiler.Program.Main(String[] args) + 0x169
         at crossgen2!<BaseAddress>+0x767205
```
this is the failing assertion: https://github.com/dotnet/runtime/blob/031a16ae6797e3ee0ebf718364a533a76b9083bc/src/libraries/Common/src/Interop/Unix/System.Native/Interop.UnixFileSystemTypes.cs#L159 and the decimal value 26 is 0x1A, which should map to apfs: https://github.com/dotnet/runtime/blob/031a16ae6797e3ee0ebf718364a533a76b9083bc/src/libraries/Common/src/Interop/Unix/System.Native/Interop.UnixFileSystemTypes.cs#L27 perhaps it was trimmed off by the linker?

Reply:
> perhaps it was trimmed off by the linker?

Yes, the enum is not public and parts of CoreLib that are not public are blocked from reflection. The concept of reflection blocking was inherited from .NET Native. It is of less importance now that our approach to reflection is different than it was in .NET Native. #72570 tracks getting rid of it. The hope is that the problem will go away eventually.

As for workarounds, we could add

```xml
<!-- Some asserts in CoreLib are not compatible with metadata blocking -->
<ItemGroup  Condition="'$(Configuration)' != 'Release'>
    <IlcArg Include="--nometadatablocking" />
</ItemGroup>
```

to crossgen2.csproj

Reply:
@MichalStrehovsky, thank you! Just curious, we are excluding `nometadatablocking` for Release because we are not running the tests with release configuration (and not because `nometadatablocking` is implicitly set for release), right?

Reply:
It may be the case that the problematic reflection constructs only affect [debug-only] assertion checks, not "actual" runtime functionality, and as we're skipping asserts in release build mode there's no reason to bloat the NativeAOT Crossgen2 app by suppressing some of the IL Linker filtering; it also makes sure that the shipping release build mode of Crossgen2 is not affected by this workaround.